src/pds/cross_report.rs

```diff
@@ -0,0 +1,404 @@
+        // requested, abort and return null report
+        for bucket in &relevant_event_selector.requested_buckets {
+            if self.already_requested_buckets.contains(bucket) {
+                debug!("Bucket {bucket:?} already requested, returning null report");
+                return Ok(PdsReport::default());
+            }
+        }
+
+        // Add the requested buckets to the already requested set
+        self.already_requested_buckets
+            .extend(&relevant_event_selector.requested_buckets);
+
+        let mut relevant_events = self.events.clone();
+        let mut unfiltered_report =
+            self.request.compute_report(&relevant_events);
+
+        // only keep the buckets that are requested
+        unfiltered_report.bin_values.retain(|bucket, _| {
+            relevant_event_selector.requested_buckets.contains(bucket)
+        });
+
+        let mut oob_filters = vec![];
+        for epoch_id in epochs {
+            let epoch_relevant_events = self.events.for_epoch(&epoch_id);
+
+            // Step 2. Compute individual loss for current epoch.
+            let individual_privacy_loss = compute_epoch_loss(
+                &self.request,
+                epoch_relevant_events,
+                &unfiltered_report,
+                num_epochs,
+            );
+
+            let filter_id =
+                FilterId::PerQuerier(epoch_id, beneficiary_uri.clone());
+            let filter_status = filter_storage
+                .try_consume(&filter_id, &individual_privacy_loss)?;
+
+            if filter_status == FilterStatus::OutOfBudget {
+                // Not enough budget, drop events without any filter
+                // consumption
+                relevant_events.drop_epoch(&epoch_id);
+
+                // Keep track of why we dropped this epoch
+                oob_filters.push(filter_id);
+            }
+        }
+
+        // Now that we've dropped OOB epochs, we can compute the final report.
+        let mut filtered_report = self.request.compute_report(&relevant_events);
+
+        // only keep the buckets that are requested
+        filtered_report.bin_values.retain(|bucket, _| {
+            relevant_event_selector.requested_buckets.contains(bucket)
+        });
+
+        let report = PdsReport {
+            filtered_report,
+            unfiltered_report,
+            oob_filters,
+        };
+        Ok(report)
+    }
+}
+
+impl<U, FS, ERR> PrivateDataServiceCore<PpaHistogramRequest<U>, FS, ERR>
+where
+    U: Uri,
+    FS: FilterStorage<
+        FilterId = FilterId<PpaEpochId, U>,
+        Budget = PureDPBudget,
+    >,
+    ERR: From<FS::Error>,
+{
+    pub fn measure_conversion(
+        &mut self,
+        request: PpaHistogramRequest<U>,
+        mut relevant_events: RelevantEvents<PpaEvent<U>>,
+    ) -> Result<AttributionObject<PpaHistogramRequest<U>>, ERR> {
+        let uris = request.report_uris();
+
+        let epochs = request.epoch_ids();
```

Comment:
Could you add something as follows, if epoch_ids has more than two elements?
```
warn!("Cross-report optimization only saves budget when requesting more than 1 epoch. We recommend using the regular API otherwise.")
```

Reply:
Added in https://github.com/columbia/pdslib/pull/95/commits/614c352e9e34d709433c1c2c40cd3e4886599477